src/cron.ts

```diff
@@ -104,22 +93,39 @@ async function scheduleCronJobs(
   escapedWorkerSchema: string,
   jobsAndIdentifiers: JobAndCronIdentifier[],
   ts: string,
+  backfilled: boolean,
 ) {
   // Note that `identifier` is guaranteed to be unique for every record
   // in `specs`.
   await pgPool.query(
     `
-      with specs as (
+      with known_crontabs as (
+
+      ),
+      specs as (
         select
           index,
           (json->>'identifier')::text as identifier,
           ((json->'job')->>'task')::text as task,
-          ((json->'job')->'payload')::json as payload,
+          (((json->'job')->'payload')::jsonb || jsonb_build_object(
```

Comment:
Casting between `json` and `jsonb` is expensive; if we can avoid it I'd rather we did.

Further, I'm pretty sure we already know the values from `known_crontabs` when scheduling the jobs, so requesting this data from the DB again should be unnecessary. We should be able to populate this `_cron` value entirely in JS.

Reply:
I think we don't know these. If we did I'd use them for sure.

Reply:
Ah yes you're right; it's done as part of the CTE afterwards. Hmmm :thinking: I may need to think about this a bit more.

Reply:
Ok, please let me know what you think, I'll pause until then.

Reply:
It's going to be a few days, I'm flat out at the moment :+1:

Reply:
We _do_ know the last executed timestamp during backfilling - we get it during `registerAndBackfillItems`: https://github.com/graphile/worker/blob/5a09a37e94f49de0d531569f6ca00d5dadd816a5/src/cron.ts#L171

It's not exactly set up for this, in particular we don't create `backfillItemsAndDates` for `unknownIdentifiers` because we don't backfill them (they're unknown), so they're `knownSince: now, lastExecuted: never`. We could however add the relevant logic to enable this without needing much (if anything?) in the way of query changes.

However, we have to keep in mind that multiple workers are quite likely to be scheduling the same cron jobs at the same time, we do this on a "first call wins" basis via `on conflict ... where` in the `locks` CTE. That said, if we didn't hit that we can assume that someone else did so we can advance the `last_execution` timestamp. So I'm pretty sure we can factor this information into `makeJobForItem` calls and not have to change the `scheduleCronJobs` query at all.

That said... <see comment on PR as a whole>